datalad/distribution/get.py

```diff
@@ -754,11 +770,18 @@ def __call__(
                         dsrec = content_by_ds.get(res['path'], set())
                         dsrec.update(res['contains'])
                         content_by_ds[res['path']] = dsrec
-                    if res.get('status', None) != 'notneeded':
-                        # all those messages on not having installed anything
-                        # are a bit pointless
-                        # "notneeded" for annex get comes below
+
+                    # we want to yield dataset results only once (not on every input path) and
+                    # in addition "notneeded" results are relevant only, if those paths were
+                    # actually requested:
+                    if res['path'] not in yielded_dataset_paths and \
+                            (res.get('status', None) != 'notneeded' or res.get('path', None) in input_paths):
                         yield res
+                        # record that we have a dataset-type result for that path already
+                        # to avoid following up with an additional directory-type result
+                        # via annex-get or just another notneeded dataset-type result as
+                        # we deal with another input path
+                        yielded_dataset_paths.append(res['path'])
```

Comment:
oi -- dejavu... I think I saw this code before... duplication, suggests needed RF of logic or coming up with a helper so just to have `if should_dataset_be_yielded(res): yield res` which would in turn memoize what needed in that `yielded_dataset_paths` or alike within `should_dataset_be_yielded`

Reply:
Agree. Duplication was there before and for now just tried to figure that out while being minimally invasive. If we agree on what's the solution - this should be RF'ed, yes.